lib/connections/service.go

```diff
@@ -391,23 +395,62 @@ func (s *Service) connect() {
 					continue
 				}
 
-				if priorityKnown && dialerFactory.Priority() >= ct.internalConn.priority {
+				prio := dialerFactory.Priority()
+
+				if priorityKnown && prio >= ct.internalConn.priority {
 					l.Debugf("Not dialing using %s as priority is less than current connection (%d >= %d)", dialerFactory, dialerFactory.Priority(), ct.internalConn.priority)
 					continue
 				}
 
 				dialer := dialerFactory.New(s.cfg, s.tlsCfg)
-				l.Debugln("dial", deviceCfg.DeviceID, uri)
 				nextDial[addr] = now.Add(dialer.RedialFrequency())
 
-				conn, err := dialer.Dial(deviceID, uri)
-				if err != nil {
-					l.Debugln("dial failed", deviceCfg.DeviceID, uri, err)
-					continue
+				// Make sure we don't add dupes, so that sort behaves the way
+				// we expect.
+				if _, ok := dialFuncsPerPriority[prio]; !ok {
+					priorities = append(priorities, prio)
 				}
 
-				s.conns <- conn
-				continue nextDevice
+				dialFuncsPerPriority[prio] = append(dialFuncsPerPriority[prio], func() (internalConn, error) {
+					l.Debugln("dial", deviceCfg.DeviceID, uri)
+					c, err := dialer.Dial(deviceID, uri)
+					l.Debugln("dial", deviceCfg.DeviceID, uri, "outcome", c, err)
+					return c, err
+				})
+			}
+
+			sort.Ints(priorities)
+
+			for _, prio := range priorities {
+				funcs, ok := dialFuncsPerPriority[prio]
+				if !ok {
+					continue
+				}
+				wg := sync.NewWaitGroup()
+				wg.Add(len(funcs))
+				res := make(chan internalConn, len(funcs))
+				for _, dialFunc := range funcs {
+					go func(df func() (internalConn, error), result chan internalConn) {
+						if c, err := df(); err == nil {
+							result <- c
+						}
+						wg.Done()
+					}(dialFunc, res)
+				}
+				wg.Wait()
```

Comment:
It seems wasteful to wait for all connections to fail before proceeding with one that succeeded?

Reply:
It would be awesome if we'd have context support in the dialer package, yet we don't :(